redis/src/main/java/org/hibernate/ogm/datastore/redis/RedisDialect.java

```diff
@@ -0,0 +1,546 @@
+import org.hibernate.ogm.model.key.spi.AssociationKeyMetadata;
+import org.hibernate.ogm.model.key.spi.EntityKey;
+import org.hibernate.ogm.model.key.spi.EntityKeyMetadata;
+import org.hibernate.ogm.model.key.spi.RowKey;
+import org.hibernate.ogm.model.spi.AssociationKind;
+import org.hibernate.ogm.model.spi.Tuple;
+import org.hibernate.ogm.options.spi.OptionsContext;
+import org.hibernate.ogm.type.spi.GridType;
+import org.hibernate.persister.entity.Lockable;
+import org.hibernate.type.Type;
+
+import com.lambdaworks.redis.KeyScanCursor;
+import com.lambdaworks.redis.RedisConnection;
+import com.lambdaworks.redis.ScanArgs;
+import com.lambdaworks.redis.protocol.LettuceCharsets;
+
+/**
+ * Stores tuples and associations inside Redis.
+ * <p/>
+ * Tuples are stored in Redis as a JSON serialization of a {@link Entity} object. Associations are stored in Redis obtained as a
+ * JSON serialization of a {@link Association} object either within the entity or external. See {@link RedisProperties#ASSOCIATIONS_STORE} on how to configure
+ * entity or external storage.
+ *
+ * @author Mark Paluch
+ */
+public class RedisDialect extends BaseGridDialect {
+
+	public static final String IDENTIFIERS = "Identifiers";
+	public static final byte[] IDENTIFIERS_AS_BYTES = toBytes( IDENTIFIERS );
+	public static final String ASSOCIATIONS = "Associations";
+
+	protected final Map<EntityStorageType, EntityStorageStrategy> entityStorageStrategies;
+
+	private final RedisDatastoreProvider provider;
+	private final RedisConnection<byte[], byte[]> connection;
+
+	private final JsonSerializationStrategy serializationStrategy = new JsonSerializationStrategy();
+
+	public RedisDialect(RedisDatastoreProvider provider) {
+		this.provider = provider;
+		this.connection = provider.getConnection();
+
+		Map<EntityStorageType, EntityStorageStrategy> strategies = new HashMap<>();
+		strategies.put( EntityStorageType.JSON, new JsonEntityStorageStrategy( serializationStrategy, connection ) );
+		strategies.put( EntityStorageType.HASH, new ExperimentalHashEntityStorageStrategy( connection ) );
+
+		this.entityStorageStrategies = Collections.unmodifiableMap( strategies );
+
+	}
+
+	/**
+	 * Redis essentially has no workable lock strategy
+	 *
+	 * @param lockable The persister for the entity to be locked.
+	 * @param lockMode The type of lock to be acquired.
+	 *
+	 * @return always null
+	 */
+	@Override
+	public LockingStrategy getLockingStrategy(Lockable lockable, LockMode lockMode) {
+		// Redis essentially has no workable lock strategy
+		return null;
+	}
+
+	@Override
+	public GridType overrideType(Type type) {
+		return serializationStrategy.overrideType( type );
+	}
+
+	@Override
+	public Tuple getTuple(EntityKey key, TupleContext tupleContext) {
+
+		Entity entity = getEntity( key, tupleContext.getOptionsContext() );
+		if ( entity != null ) {
+			return new Tuple( new RedisTupleSnapshot( entity.getProperties() ) );
+		}
+		else if ( isInTheQueue( key, tupleContext ) ) {
```

Comment:
This code is c'n'p from MongoDB because I assumed it was needed. I can remove the queueing part. Can you explain to me why you don't like how it's done?